primitives/runtime/src/offchain/rpc.rs

```diff
@@ -0,0 +1,353 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A high-level helpers for making RPC calls from Offchain Workers.
+//!
+//! Example:
+//! ```rust,no_run
+//! use sp_runtime::offchain::rpc::{Request, Error};
+//! use sp_runtime::offchain::http;
+//! use sp_std::{vec::Vec, str};
+//!
+//! let request = Request {
+//! 	jsonrpc: "2.0",
+//! 	id: 1,
+//! 	method: "chain_getFinalizedHead",
+//! 	params: Vec::new(),
+//! 	timeout: 3_000,
+//! 	url: "http://localhost:9933"
+//! };
+//!
+//! let rpc_response = request.send();
+//!
+//! match rpc_response {
+//! 	Ok(response) => {
+//! 		if !response.result.is_empty() {
+//! 			log::info!("Rpc call result: {:?}", str::from_utf8(&response.result).unwrap());
+//! 		} else {
+//! 			log::error!("Rpc call error: code: {:?}, message: {:?}",
+//! 							response.error.code,
+//! 							str::from_utf8(&response.error.message).unwrap()
+//! 						);
+//! 		}
+//! 	},
+//! 	Err(e) => {
+//! 		log::error!("Rpc call error: {:?}", e);
+//! 		assert_eq!(Error::Http(http::Error::DeadlineReached), e);
+//! 	}
+//! }
+//! ```
+
+use serde::{Deserialize, Deserializer};
+use crate::{RuntimeDebug, offchain::Duration};
+use serde_json::json;
+use sp_std::{vec::Vec, str};
+use super::http;
+
+/// A rpc call error
+#[derive(Clone, PartialEq, Eq, RuntimeDebug)]
+pub enum Error {
+	/// Response can not be deserialized
+	Deserializing,
+	/// Http error
+	Http(http::Error),
+}
+
+/// A rpc result type
+pub type RpcResult = Result<Response, Error>;
+
+/// A rpc call error field
+#[derive(Deserialize, Default, RuntimeDebug)]
+pub struct RpcError {
+	/// error code
+	pub code: i32,
+	/// error message
+	#[serde(deserialize_with = "de_string_to_bytes")]
+	pub message: Vec<u8>,
+}
+
+/// A rpc call response
+#[derive(Deserialize, Default, RuntimeDebug)]
+pub struct Response {
+	/// rpc response jsonrpc
+	#[serde(deserialize_with = "de_string_to_bytes")]
+	pub jsonrpc: Vec<u8>,
+	/// rpc response result
+	#[serde(default)]
+	#[serde(deserialize_with = "de_string_to_bytes")]
+	pub result: Vec<u8>,
+	/// rpc response error struct
+	#[serde(default)]
+	pub error: RpcError,
+	/// rpc response id
+	pub id: u32,
+}
+
+/// A rpc call request
+pub struct Request<'a> {
```

Comment:
```suggestion
#[derive(Serialize, RuntimeDebug)]
pub struct Request<'a> {
```

1. Any reason not to use `Serialize` from serde instead of `json!` macro?
2. Are there any issue with `RuntimeDebug`?

Reply:
- Not applicable anymore because of previous comment:

  > @NachoPal so sorry for going back and worth, but can we avoid using derive and only use serde_json to serialize/deserialize the structs (I guess it means using json! macro and maybe deseiralizing to Value and then extracting the fields).

- `RuntimeDebug` added